### PR TITLE
Hyperblock processor

### PR DESCRIPTION
**Pull request by ssd04**
Added hyper outport block processor components:
- initial implementation for blocks pool handler
- outport block data aggregator
- firehose publisher

### CODE REVIEW

File: process/dataProcessor.go [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+func (dp *dataProcessor) saveBlock(marshalledData []byte) error {
+	outportBlock := &outport.OutportBlock{}
+	err := dp.marshaller.Unmarshal(outportBlock, marshalledData)
+	if err != nil {
+		return err
+	}
+
+	if outportBlock == nil || outportBlock.BlockData == nil {
```
[the remark this answers is not in the document]
Comment (ssd04) [reply process/dataProcessor.go:69]: yes, if `marshalledData` is nil, the marshaller will return nil on Unmarshall

[the remark this answers is not in the document]
Comment (cristure) [reply process/dataProcessor.go:69]: right. missed that, thanks for the clarification.

File: process/dataAggregator.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+	return &dataAggregator{
+		blocksPool: blocksPool,
+	}, nil
+}
+
+// ProcessHyperBlock will process meta outport block. It will try to fetch and aggregate
+// notarized shards data
+func (da *dataAggregator) ProcessHyperBlock(outportBlock *outport.OutportBlock) (*data.HyperOutportBlock, error) {
+	hyperOutportBlock := &data.HyperOutportBlock{}
+	hyperOutportBlock.MetaOutportBlock = outportBlock
```
Comment (AdoAdoAdo) [inline process/dataAggregator.go:32]: maybe some check that the received outport block is indeed a metablock, e.g. check the ShardID?

File: go.mod [hunk 3]
```diff
@@ -3,9 +3,10 @@ module github.com/multiversx/mx-chain-ws-connector-template-go
 go 1.20
 
 require (
-	github.com/multiversx/mx-chain-communication-go v1.0.12
-	github.com/multiversx/mx-chain-core-go v1.2.18
-	github.com/multiversx/mx-chain-logger-go v1.0.13
+	github.com/multiversx/mx-chain-communication-go v1.0.13-0.20240126121117-627adccf10ad
```
[the remark this answers is not in the document]
Comment (ssd04) [reply go.mod]: latest storage version does not work with core 1.2.18, we might need to have a new storage version

File: process/dataProcessor.go [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+// NewDataProcessor creates a data processor able to receive data from a ws outport driver and handle blocks
+func NewDataProcessor(
+	publisher Publisher,
+	marshaller marshal.Marshalizer,
+	blocksPool BlocksPool,
+	dataAggregator DataAggregator,
+) (DataProcessor, error) {
+	if publisher == nil {
```
[the remark this answers is not in the document]
Comment (ssd04) [reply process/dataProcessor.go:25]: updated to use check.IfNil

File: process/dataProcessor.go [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+	err := dp.blocksPool.PutBlock(blockHash, outportBlock)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Close will close the internal writer
+func (dp *dataProcessor) Close() error {
```
[the remark this answers is not in the document]
Comment (ssd04) [reply process/dataProcessor.go:108]: this is done on next PR :+1:

File: process/dataProcessor_test.go [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+	require.Nil(t, dp.ProcessPayload([]byte("payload"), outportcore.TopicSaveValidatorsRating, 1))
+	require.Nil(t, dp.ProcessPayload([]byte("payload"), outportcore.TopicSaveValidatorsPubKeys, 1))
+	require.Nil(t, dp.ProcessPayload([]byte("payload"), outportcore.TopicSaveAccounts, 1))
+	require.Nil(t, dp.ProcessPayload([]byte("payload"), outportcore.TopicFinalizedBlock, 1))
+}
+
+func TestDataProcessor_ProcessPayload(t *testing.T) {
+	t.Parallel()
+
+	t.Run("nil outport block, should return error", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (ssd04) [reply process/dataProcessor_test.go:142]: done

File: process/dataAggregator.go (hunk 2 above)
Comment (ssd04) [reply process/dataAggregator.go:32]: added shard id check

File: go.mod (hunk 3 above)
[the remark this answers is not in the document]
Comment (ssd04) [reply go.mod]: in the meantime, we have storage v1.0.15 which works well with latest version; updated storage version

**Review by cristure — APPROVED** [review]

**Pull request merged**